From 8c64327874cbcfd18b802ceee876648d95024f74 Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Sun, 29 Dec 2019 14:39:50 +0800
Subject: [PATCH 143/151] driver: edac: create an interface to check ecc
 feature status

Create an interface, SMC instruction is used to check ecc status by
reading ecc ctrl1 register in EL3 level.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
Integrated-by: Jun Zhang <jun.zhang@windriver.com>
---
 drivers/edac/altera_edac.c | 20 ++++++++++++++++++++
 drivers/edac/altera_edac.h |  3 +++
 2 files changed, 23 insertions(+)

diff --git a/drivers/edac/altera_edac.c b/drivers/edac/altera_edac.c
index 1b4f6df82eec..be89b4db9825 100644
--- a/drivers/edac/altera_edac.c
+++ b/drivers/edac/altera_edac.c
@@ -276,6 +276,22 @@ static int a10_unmask_irq(struct platform_device *pdev, u32 mask)
 	return ret;
 }
 
+static bool s10_is_ecc_enable(void)
+{
+	struct arm_smccc_res result;
+
+	arm_smccc_smc(INTEL_SIP_SMC_REG_READ, S10_ECCCTRL1_OFST, 0, 0, 0,
+		      0, 0, 0, &result);
+
+	if (A10_ECCCTRL1_ECC_EN & result.a1 == A10_ECCCTRL1_ECC_EN)
+		return true;
+	else {
+		edac_printk(KERN_ERR, EDAC_MC,
+			    "No ECC/ECC disabled [0x%08X]\n", result.a1);
+		return false;
+	}
+}
+
 static int altr_sdram_probe(struct platform_device *pdev)
 {
 	const struct of_device_id *id;
@@ -289,6 +305,10 @@ static int altr_sdram_probe(struct platform_device *pdev)
 	int irq, irq2, res = 0;
 	unsigned long mem_size, irqflags = 0;
 
+	if (of_machine_is_compatible("altr,socfpga-stratix10"))
+		if (!s10_is_ecc_enable())
+			return -ENODEV;
+
 	id = of_match_device(altr_sdram_ctrl_of_match, &pdev->dev);
 	if (!id)
 		return -ENODEV;
diff --git a/drivers/edac/altera_edac.h b/drivers/edac/altera_edac.h
index ef2e2448ddc7..7ee43be08f6e 100644
--- a/drivers/edac/altera_edac.h
+++ b/drivers/edac/altera_edac.h
@@ -354,6 +354,9 @@ struct altr_sdram_mc_data {
 #define S10_COLD_RESET_MASK               0x30002
 #define S10_WARM_RESET_WFI_FLAG           BIT(31)
 
+/* SDRAM Controller EccCtrl Register */
+#define S10_ECCCTRL1_OFST          	0xF8011100
+
 struct altr_edac_device_dev;
 
 struct edac_device_prv_data {
-- 
2.26.1

